src/main/scala/devices/spi/SPIPhysical.scala

```diff
@@ -16,9 +17,26 @@ object SPIMicroOp {
   def Delay    = UInt(1, 1)
 }
 
+//Coarse delay is the number of system-clock cycles that can be added
+//as a phase difference between sent and received SPI data
+//Fine delay is the fine-grain delay that can be added as a phase 
+//difference between send and received SPI data.
+//Fine delay is typically achieved through foundry specific delay buffers
+class SPIExtraDelay(c: SPIParamsBase) extends SPIBundle(c) {
+  val coarse = UInt(width = c.divisorBits)
+  val fine = UInt(width = c.fineDelayBits)
+}
+//Sample delay reflects minimum sequential delay that exists between 
```

Comment:
Thanks for these comments. Both coarse delay and sample delay are SW-writable registers, right? And I see that somewhere below they are added together to make the actual delay. So what is the difference? Why do we have two?

Reply:
What I mean is, it seems we always only ever use the `totalDelay`, so why do we need both sampleDelay and extraDelay?

Reply:
sampleDelay is a necessary delay (indicating both distance and/or sychronization delays between a slave and the master. Total delay cannot be less than sampleDelay. However coarseDealy can be 0 ( in which case TotalDelay equals sampleDelay).

Reply:
Both parameters represent different things and hence are intentionally kept separate.

Reply:
Maybe I'm confused, but isn't sampleDelay still a SW-writable register? If sampleDelay is necessary, why can SW write the register?

Reply:
It is SW writeable. The coarseDelay is to improve speed, sampleDelay is to correctly sample.